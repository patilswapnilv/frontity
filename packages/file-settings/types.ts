/**
 * Abstract base type used to make the {@link Multi} and {@link Normalized} types.
 */
type Base = {
  /**
   * Field to distinguish between different sites.
   */
  match?: string[];

  /**
   * The mode of this site.
   *
   * @example "amp"
   *
   * @defaultValue "default"
   */
  mode?: string;

  /**
   * Global state.
   */
  state?: Record<string, unknown>;
};

/**
 * The base type of a Frontity package.
 */
export type Package = {
  /**
   * Package name.
   */
  name: string;

  /**
   * Whether the package should be active or not.
   */
  active?: boolean;

  /**
   * Package state.
   */
  state?: Record<string, unknown>;
};

/**
 * Package that has been normalized.
 */
type NormalizedPackage = Required<Package>;

/**
 * Multi-site package.
 */
type Multi = Base & {
  /**
   * Site name.
   */
  name: string;
};

/**
 * Represents normalized single site settings.
 */
type Normalized<T> = Base & {
  /**
   * Mode of the site.
   */
  mode: string;

  /**
   * Site state.
   */
  state: Record<string, unknown>; // Default: {}

  /**
   * Packages of the site.
   */
  packages: T[];
};

/**
 * Normalized site settings.
 */
export type NormalizedSettings<T = NormalizedPackage> = Normalized<T> & Multi;

/**
 * A Frontity site.
 */
export type Site = {
  /**
   * Site name.
   */
  name: string;

  /**
   * The mode of this site.
   */
  mode: string;

  /**
   * Site packages.
   */
  packages: string[];
};
