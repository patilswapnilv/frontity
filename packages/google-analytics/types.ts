import * as React from "react";
import { Action, MergePackages, Package, Frontity } from "frontity/types";
import Analytics, {
  Pageview,
  Event,
  AmpConfig,
} from "@frontity/analytics/types";

declare global {
  /**
   * Extended with Google Analytics global variables.
   */
  interface Window {
    /**
     * Google Analytics command queue function.
     */
    gtag: (...args: any) => any;
  }
}

/**
 * Google Analytics event that `actions.googleAnalytics.event` receives as
 * argument.
 *
 * @example
 * ```
 * {
 *   name: "Click",
 *   payload: {
 *     category: "Video",
 *     label: "Featured Media",
 *     value: 42,
 *   },
 * }
 * ```
 */
interface GoogleAnalyticsEvent extends Event {
  /**
   * Event name.
   *
   * The value of this property is mapped to the [`eventAction`](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#eventAction)
   * field of `analytics.js` events.
   *
   * @example "Click"
   */
  name: string;

  /**
   * Event payload.
   *
   * @example
   * ```
   * {
   *   category: "Video",
   *   label: "Featured Media",
   *   value: 42
   * }
   * ```
   */
  payload: {
    /**
     * Event category.
     *
     * The value of this property is mapped to the [`eventCategory`](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#eventCategory)
     * field of `analytics.js` events.
     *
     * @example "Video"
     */
    category: string;

    /**
     * Event label.
     *
     * The value of this property is mapped to the [`eventLabel`](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#eventLabel)
     * field of `analytics.js` events.
     *
     * @example "Featured Media"
     */
    label?: string;

    /**
     * Event value.
     *
     * The value of this property is mapped to the [`eventValue`](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#eventValue)
     * field of `analytics.js` events.
     *
     * @example 42
     */
    value?: number;

    /**
     * Any other property specified in [`analytics.js` field reference](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference).
     */
    [key: string]: any;
  };
}

/**
 * AMP config with GoogleAnalytics-specific variables.
 */
export interface GoogleAnalyticsAmpConfig extends AmpConfig {
  /**
   * AmpConfig vars.
   */
  vars?: AmpConfig["vars"] & {
    /**
     * Primary tracking ID.
     */
    gtag_id?: string;

    /**
     * Object containing all tracking IDs and their configuration.
     */
    config?: Record<string, unknown>;
  };
}

/**
 * Analytics package to use Google Analytics with Frontity.
 */
interface GoogleAnalytics extends Package {
  /**
   * Package name.
   */
  name: "@frontity/google-analytics";

  /**
   * Root elements exposed by this package.
   */
  roots: Analytics["roots"] & {
    /**
     * Google Analytics root element.
     */
    googleAnalytics: React.ElementType;
  };

  /**
   * The state exposed by this package.
   */
  state: Analytics["state"] & {
    /**
     * State properties for the Google analytics package.
     */
    googleAnalytics: {
      /**
       * Tracking ID associated to a Google Analytics account.
       */
      trackingId?: string;

      /**
       * List of tracking IDs associated to a Google Analytics account.
       */
      trackingIds?: string[];

      /**
       * JSON configuration object to include inside the `<amp-analytics>` tags,
       * useful to add variables and triggers. Note that this props is only used
       * when the `@frontity/amp` package is installed.
       *
       * See
       * https://developers.google.com/analytics/devguides/collection/amp-analytics/
       * for more info about how to define this config object.
       */
      ampConfig?: GoogleAnalyticsAmpConfig;
    };
  };

  /**
   * The actions exposed by this package.
   */
  actions: Analytics["actions"] & {
    /**
     * Actions from the Google analytics package.
     */
    googleAnalytics: {
      /**
       * Send a pageview to all traking IDs defined in
       * either `state.googleAnalytics.trackingId` or
       * `state.googleAnalytics.trackingIds`.
       *
       * @remarks
       * This action is called automatically if
       * `state.analytics.pageviews.googleAnalytics` is set to `true`.
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<Packages, Pageview>;

      /**
       * Send an event to all traking IDs defined in
       * either `state.googleAnalytics.trackingId` or
       * `state.googleAnalytics.trackingIds`.
       *
       * @remarks
       * This action is called automatically when calling
       * `actions.analytics.event()` and
       * `state.analytics.pageviews.googleAnalytics` is set to `true`.
       *
       * @example
       * ```
       * actions.googleAnalytics.event({
       *   name: "Click",
       *   payload: {
       *     category: "Video",
       *     label: "Featured Media",
       *     value: 42,
       *   },
       * });
       * ```
       *
       * @param event - Object of type {@link GoogleAnalyticsEvent}.
       */
      event: Action<Packages, GoogleAnalyticsEvent>;
    };
  };
}

/**
 * Package types used internally by GoogleAnalytics.
 */
export type Packages = MergePackages<Frontity, GoogleAnalytics, Analytics>;

export default GoogleAnalytics;
