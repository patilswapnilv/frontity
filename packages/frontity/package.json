{
  "name": "frontity",
  "version": "0.1.4",
  "description": "Frontity cli and entry point to other packages",
  "keywords": [
    "frontity"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "main": "src/index.js",
  "bin": {
    "frontity": "build/index.js"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "start": "ts-node src/index.ts"
  },
  "devDependencies": {
    "@types/jest": "^24.0.11",
    "@types/node": "^11.13.2",
    "jest": "^24.7.1",
    "ts-node": "^8.0.3",
    "typescript": "^3.4.2"
  },
  "dependencies": {
    "commander": "^2.20.0"
  }
}
