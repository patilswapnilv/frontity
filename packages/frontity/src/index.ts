import * as program from "commander";
import create from "./create";
import { version } from "../package.json";

// Sets the version and the description of the program.
program.version(version).description("Frontity CLI");

// Register the `create` command.
// - Takes an optional parameter called `name`.
// - Creates a new Frontity app.
program
  .command("create [name]")
  .description("Creates a new Frontity app.")
  .action(create);

program.parse(process.argv);
