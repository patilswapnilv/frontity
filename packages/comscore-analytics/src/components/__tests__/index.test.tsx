/**
 * @jest-environment node
 */

import TestRenderer from "react-test-renderer";
import { HelmetProvider } from "frontity";
import { State } from "frontity/types";
import { FilledContext } from "react-helmet-async";
import { Root as ComscoreAnalytics } from "..";
import { Packages } from "../../../types";

const getState = (): State<Packages> => ({
  frontity: {},
  analytics: {
    pageviews: { comscoreAnalytics: true },
    events: {},
  },
  comscoreAnalytics: {},
});

describe("Comscore Analytics Component", () => {
  test("should render script for all tracking IDs", () => {
    const state = getState();

    state.comscoreAnalytics.trackingIds = ["111111", "222222"];

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(head.noscript.toComponent()).toMatchSnapshot();
  });

  test("should render a script for a single tracking ID", () => {
    const state = getState();

    state.comscoreAnalytics.trackingId = "333333";

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(head.noscript.toComponent()).toMatchSnapshot();
  });

  test("should not render if no tracking IDs are specified", () => {
    const state = getState();

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(head.noscript.toComponent()).toMatchSnapshot();
  });
});

describe("Comscore Analytics Component (AMP)", () => {
  test("should render script for all tracking IDs", () => {
    const state = getState();
    state.frontity.mode = "amp";
    state.comscoreAnalytics.trackingIds = ["111111", "222222"];

    const helmetContext = {};
    const app = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(app).toMatchSnapshot();
  });

  test("should render a script for a single tracking ID", () => {
    const state = getState();
    state.frontity.mode = "amp";
    state.comscoreAnalytics.trackingId = "333333";

    const helmetContext = {};
    const app = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(app).toMatchSnapshot();
  });

  test("should not render if no tracking IDs are specified", () => {
    const state = getState();
    state.frontity.mode = "amp";

    const helmetContext = {};
    const app = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} libraries={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toComponent()).toMatchSnapshot();
    expect(app).toMatchSnapshot();
  });
});
