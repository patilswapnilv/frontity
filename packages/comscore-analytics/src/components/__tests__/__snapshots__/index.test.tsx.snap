// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comscore Analytics Component (AMP) should not render if no tracking IDs are specified 1`] = `Array []`;

exports[`Comscore Analytics Component (AMP) should not render if no tracking IDs are specified 2`] = `null`;

exports[`Comscore Analytics Component (AMP) should render a script for a single tracking ID 1`] = `Array []`;

exports[`Comscore Analytics Component (AMP) should render a script for a single tracking ID 2`] = `
<amp-analytics
  type="comscore"
>
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"vars\\":{\\"c2\\":\\"333333\\"}}",
      }
    }
    type="application/json"
  />
</amp-analytics>
`;

exports[`Comscore Analytics Component (AMP) should render script for all tracking IDs 1`] = `Array []`;

exports[`Comscore Analytics Component (AMP) should render script for all tracking IDs 2`] = `
Array [
  <amp-analytics
    type="comscore"
  >
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"vars\\":{\\"c2\\":\\"111111\\"}}",
        }
      }
      type="application/json"
    />
  </amp-analytics>,
  <amp-analytics
    type="comscore"
  >
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"vars\\":{\\"c2\\":\\"222222\\"}}",
        }
      }
      type="application/json"
    />
  </amp-analytics>,
]
`;

exports[`Comscore Analytics Component should not render if no tracking IDs are specified 1`] = `Array []`;

exports[`Comscore Analytics Component should not render if no tracking IDs are specified 2`] = `Array []`;

exports[`Comscore Analytics Component should render a script for a single tracking ID 1`] = `
Array [
  <script
    async={true}
    data-rh={true}
    src="https://sb.scorecardresearch.com/beacon.js"
  />,
]
`;

exports[`Comscore Analytics Component should render a script for a single tracking ID 2`] = `
Array [
  <noscript
    dangerouslySetInnerHTML={
      Object {
        "__html": "<img alt=\\"comscore\\" src=\\"https://sb.scorecardresearch.com/p?c1=2&c2=333333&cv=2.0&cj=1\\" />",
      }
    }
    data-rh={true}
  />,
]
`;

exports[`Comscore Analytics Component should render script for all tracking IDs 1`] = `
Array [
  <script
    async={true}
    data-rh={true}
    src="https://sb.scorecardresearch.com/beacon.js"
  />,
]
`;

exports[`Comscore Analytics Component should render script for all tracking IDs 2`] = `
Array [
  <noscript
    dangerouslySetInnerHTML={
      Object {
        "__html": "<img alt=\\"comscore\\" src=\\"https://sb.scorecardresearch.com/p?c1=2&c2=111111&cv=2.0&cj=1\\" />",
      }
    }
    data-rh={true}
  />,
  <noscript
    dangerouslySetInnerHTML={
      Object {
        "__html": "<img alt=\\"comscore\\" src=\\"https://sb.scorecardresearch.com/p?c1=2&c2=222222&cv=2.0&cj=1\\" />",
      }
    }
    data-rh={true}
  />,
]
`;
